dash/pages/cohorts.py

```diff
@@ -0,0 +1,199 @@
+import logging
+
+import dash
+import dash_bootstrap_components as dbc
+from dash import dcc, html, Output, Input, ALL, callback, MATCH, State, get_app, dash_table
+from dotwiz import DotWiz
+from figures.histogram import histogram_selects, histogram_sliders
+from inflection import titleize, pluralize
+from models.file import get_file_histograms, get_files
+from models.observation import get_observation_histograms, get_patients
+import json
+import collections
+
+logger = logging.getLogger('dash')
+
+dash.register_page(__name__)
+
+
+def tree_dict():
+    """A recursive default dict."""
+    return collections.defaultdict(tree_dict)
+
+
+def build_filters(values, ids):
+    """Create a filter from selected values and ids."""
+    filters = tree_dict()
+    for value, id_ in zip(values, ids):
+        if not value:
+            continue
+        entity, parameter = id_["index"].split('-')
+        if 'AND' not in filters[entity]['filter']:
+            filters[entity]['filter']['AND'] = []
+        filters[entity]['filter']['AND'].append({'IN': {parameter: value}})
+    return filters
+
+
+@callback(
+    Output('query-builder', 'children'),
+    Input({'type': 'query-parameter', 'index': ALL}, 'value'),
+    Input({'type': 'query-parameter', 'index': ALL}, 'id')
+)
+def display_filters(values, ids):
+    """Build graphql filter."""
+    filters = build_filters(values, ids)
+    return json.dumps(filters, indent=4)
+
+
+@callback(
+    Output('histogram-data', 'data'),
+    Input({'type': 'query-parameter', 'index': ALL}, 'value'),
+    Input({'type': 'query-parameter', 'index': ALL}, 'id')
+)
+def update_counters(values, ids):
```

Comment:
so I guess the callback functionality attaches to some sort of external button since it is  not directly called  anywhere  within this commit?